File: Medium/0011-Container-with-most-water/python/container1.py
class Solution:
    def maxArea(self, height):
        # Start with two pointers at the extremes of the array
        left = 0
        right = len(height) - 1

        # This variable stores the best area found so far
        max_water = 0

        # Continue until the two pointers meet
        while left < right:
            # The width of the container is the distance between the two lines
            width = right - left

            # The height of the container is limited by the shorter line
            current_height = min(height[left], height[right])

            # Compute the area formed by the two lines
            current_water = width * current_height

            # Update the maximum area if the current one is better
            if current_water > max_water:
                max_water = current_water

            # Move the pointer pointing to the shorter line
            # because only this can potentially lead to a taller boundary
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        # Return the maximum amount of water found
        return max_water

File: Medium/0011-Container-with-most-water/cpp/container1.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        // Start with two pointers at the extremes of the array
        int left = 0;
        int right = static_cast<int>(height.size()) - 1;

        // This variable stores the maximum area found so far
        int maxWater = 0;

        // Continue until the two pointers meet
        while (left < right) {
            // The width is the distance between the two lines
            int width = right - left;

            // The container height is limited by the shorter line
            int currentHeight = min(height[left], height[right]);

            // Compute the current area
            int currentWater = width * currentHeight;

            // Update the maximum area if needed
            maxWater = max(maxWater, currentWater);

            // Move the pointer corresponding to the shorter line
            // because only that move can potentially improve the result
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        // Return the best area found
        return maxWater;
    }
};
